test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs

```diff
@@ -0,0 +1,456 @@
+        private readonly RequestInfo requestInfo;
+        private readonly Serializer serializer;
+        private readonly HeaderCollection headers;
+        private BulkUpdateGraph bulkUpdateGraph;
+
+        public BulkUpdateE2ETests()
+        {
+            this.context = new Container(new Uri("http://localhost:8000"));
+            this.requestInfo = new RequestInfo(context);
+            this.serializer = new Serializer(this.requestInfo);
+            this.headers = new HeaderCollection();
+            bulkUpdateGraph = new BulkUpdateGraph(this.requestInfo);
+        }
+
+        [Fact]
+        public void CallingBulkUpdate_WithNullArguments_ShouldThrowAnException()
+        {
+            Assert.Throws<ArgumentException>(() => this.context.BulkUpdate<Person>(null));
+        }
+
+        [Fact]
+        public void BulkUpdateAnEntry_WithOneLevelOfNesting_ReturnsOne_OperationResponse()
+        {
+            var expectedResponse = "{\"@context\":\"http://localhost:8000/$metadata#Persons/$delta\",\"value\":[{\"ID\":100,\"Name\":\"Bing\",\"Cars@delta\":[{\"ID\":\"1001\"}]}]}";
+
+            SetupContextWithRequestPipelineForSaving(
+                this.context,
+                expectedResponse);
+
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Bing",
+            };
+
+            var car = new Car { ID = 1001 };
+            this.context.AttachTo("Persons", person);
+            this.context.AddRelatedObject(person, "Cars", car);
+
+            DataServiceResponse response = this.context.BulkUpdate<Person>(person);
+ 
+            Assert.Single(response);
+            Assert.Single(response.Single().NestedResponse);
+
+            var changeOperationResponse = response.First() as ChangeOperationResponse;
+            EntityDescriptor entityDescriptor = changeOperationResponse.Descriptor as EntityDescriptor;
+            var returnedPerson = entityDescriptor.Entity as Person;
+
+            Assert.Equal("Bing", returnedPerson.Name);
+        }
+
+        [Fact]
+        public void BulkUpdateTwoTopLevelObjects_ReturnsTwo_OperationResponses()
+        {
+            var expectedResponse = "{\"@context\":\"http://localhost:8000/$metadata#Persons/$delta\",\"value\":[{\"ID\":100,\"Name\":\"Bing\"},{\"ID\":200,\"Name\":\"Edge\"}]}";
+
+            SetupContextWithRequestPipelineForSaving(
+                this.context,
+                expectedResponse);
+
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Bing",
+            };
+
+            var person2 = new Person
+            {
+                ID = 200,
+                Name = "Edge"
+            };
+
+            this.context.AttachTo("Persons", person);
+            this.context.AttachTo("Persons", person2);
+
+            DataServiceResponse response = this.context.BulkUpdate(person, person2);
+
+            Assert.Equal(2, response.Count());

```

Comment:
Would be good to assert the properties as well just to be sure it returns the right thing.